CASMPET-7619/CASMPET-7620/CASMPET-7621/CASMPET-7630/CASMPET-7632: Create/fix procedures for changing iSCSI workers (#6132)
Cray-HPE/docs-csm: operations/iscsi_sbps/iscsi_workers.md

> # Operational procedure for changing iSCSI worker nodes

Reviewer: If we're calling this feature "Selective iSCSI Node Personalization", then this header should probably change to something like:
```suggestion
# Adjusting Selective iSCSI Node Personalization
```

And the name of the file should match the header. So if you follow my suggestion, the file should be named `Adjusting _Selective_iSCSI_Node_Personalization.md`

> 

Reviewer: Need to add some section headers to this page.
```suggestion

## Overview

```

Reviewer: I see that this header was added, but you also need to add more later in the page, to break up the page into logical sections.

Author: with the latest changes to the doc, does this hold good still ?

Reviewer: Yes, but I'll mark this issue resolved, because this is covered as part of [this new comment](https://github.com/Cray-HPE/docs-csm/pull/6132/files#r2223447944)

> In CSM 1.6.0, all worker nodes were configured as iSCSI SBPS targets by default when CSM is installed/upgraded.

Reviewer: This is true for not just 1.6.0, I believe. And it happened not only at install/upgrade time, but any time the NCNs had personalization run (this is why they could add a worker NCN and it would still be configured and enabled for iSCSI, once that ran).

I also think what we really want to say here is that they were all enabled for iSCSI, right? Because even in CSM 1.7, they're all configured for it, but then it is possible to control which ones are actually enabled. At least, that is what I understood from our call with @dborman-hpe the other day. If "enabled" and "configured" aren't the right words, we should figure out which ones are, so we can be consistent and clear in our docs about this.

```suggestion
In CSM 1.6, all worker nodes are configured and enabled as iSCSI SBPS targets during [Management Node Personalization](Management_Node_Personalization.md).
```

Author: I think we will say 'worker node personalization' instead of 'Management Node Personalization' ? and reference the link: https://github.com/Cray-HPE/docs-csm/blob/release/1.7/operations/configuration_management/iSCSI_SBPS_Node_Personalization.md

Reviewer: I would assume that page (the iSCSI SBPS node personalization one) is going to go away in CSM 1.7, since it's no longer necessary, right?

Reviewer: I've opened [CASMPET-7632](https://jira-pro.it.hpe.com:8443/browse/CASMPET-7632) for the removal of that page. So I think my suggestion here still is correct.

Author: This was about  manual way of doing node personalization. Node personalization happens during install/upgrade time during bootprep time. But we had retained the manual way of doing node personalization post install/upgrade. Let me check within the team and getback.

Reviewer: They can now manually do it using the script I created.

Reviewer: And even if they have the option to do it manually, the normal case is that worker nodes will be configured during the automatic node personalization, right?

Author: Yes.

Reviewer: Then I think my edited link makes sense. Since this text has moved to a different page, I'll mark this comment resolved, and add this edit as part of [this comment](https://github.com/Cray-HPE/docs-csm/pull/6132/files#r2223477470)

> In CSM 1.6.0, all worker nodes were configured as iSCSI SBPS targets by default when CSM is installed/upgraded.

Reviewer: This section is giving good information, but if this page is intended to be for the operational procedure, then it feels like this information belongs on a different page. I would think you'd do something like the following:

* `Selective_iSCSI_Node_Personalization.md` - This page would contain the high level overview like you have here, and would include links to the various other related pages,
* `HSM_iSCSI_Group.md` - Details on the group, including how to create it, how to add or remove members, how to delete it. Although you only need to give a few examples, and otherwise point them to the actual HSM Groups documentation for more details on those operations.
* `Adjusting Selective iSCSI Node Personalization.md` - The page that gives the operational procedure for making changes during regular operations (i.e. not installs/upgrades), This page wouldn't need to give the explicit steps for changing the group -- that would just be a pointer to the relevant sections in `HSM_iSCSI_Group.md`. You would just have one of the steps be something like:

   ```
   1. Add or remove worker NCN xnames to the HSM group, or delete the group entirely.
      See <link> for details on how to do this.
   ```

In CSM 1.7.0, selective node personalization is supported where user/admin can limit/configure the required worker
nodes as iSCSI SBPS targets. This is done by creating HSM group named `iscsi_worker` and add worker node xnames to
this group which are required to be configured as iSCSI targets. See ‘Steps to create HSM group’ section of [HSM groups](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/operations/iscsi_sbps/HSM_groups_iscsi.md)

If user/admin is not intended to limit the worker nodes for iSCSI SBPS i.e no `iscsi_worker` group is created, then all

> worker nodes will be configured as iSCSI targets. If the user does not want to configure any of the worker nodes as iSCSI targets, then the HSM group `iscsi_worker` can be kept empty which is not suggestible as no compute nodes can be booted.

Reviewer: ```suggestion
In CSM 1.7.0, selective iSCSI node personalization is introduced. All worker nodes are still **configured** for iSCSI, but
selective node personalization gives administrators control over which workers are enabled as iSCSI SBPS targets.
The default behavior is still the same as in CSM 1.6, so if no action is taken to use this feature, then all worker nodes
will continue to be enabled as iSCSI targets.

Selective iSCSI node personalization uses an HSM group named `iscsi_worker` as the mechanism for an administrator to
specify which workers should be enabled as iSCSI targets. If this group does not exist, then all worker nodes will be
enabled as iSCSI targets. While it is technically possible to create this group and leave it empty, this will mean that
no worker nodes will be enabled as iSCSI targets, which in turn will mean that no managed nodes will be able to boot.
Therefore CSM generally treats it as an error if the HSM group exists but it contains no worker nodes.

For more details on creating this group, see [Steps to create HSM group](HSM_groups_iscsi.md#steps-to-create-hsm-group).
```

> This group needs to be created at certain stages during CSM install/upgrade:

Reviewer: This is not true now that CASMPET-7616 has merged, right? If they don't create the group, it'll just use all of the workers.
But isn't this page about giving the procedure for making changes operationally, not during an install or upgrade? If so, I'd just make that clear at the top of the page, and then there's no need to discuss the install/upgrade stuff here, right? It seems like it'll just confuse the issue for people who are coming here to do the procedure.

Author: Agree, have removed this text.

During fresh install of CSM 1.7.0: `iscsi_worker` HSM group needs to be created before the `Configure management nodes with CFS` step of [CSM 1.7.0 installation](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/install/configure_administrative_access.md#8-configure-management-nodes-with-cfs)

During upgrade of CSM 1.7.0: `iscsi_worker` HSM group needs to be created before ‘management nodes rollout’ step during CSM upgrade using IUF. See [CSM 1.7.0 Upgrade](https://github.com/Cray-HPE/docs-csm/blob/ae9482ae85bc2eede3ac61079ffed5a00fb0a126/operations/iuf/workflows/management_rollout.md)

> Post install/upgrade if it is required to add any new worker node to the cluster or need to remove an existing worker node from the cluster, then see the [add/remove worker node](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/operations/node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md#add-worker-storage-or-master-ncns).

Reviewer: This reminds me that the procedure for adding and removing worker nodes is another place we need to add a callout about this feature. They should make their changes to the HSM group just before they actually add or remove the workers, so that the next time CFS runs, it will do the right thing.

Author: Good point, so we shall skip this documentation here now?

Reviewer: Sure. I've created [CASMPET-7631](https://jira-pro.it.hpe.com:8443/browse/CASMPET-7631) to update the add/remove worker node procedure for this.

Post this, if it is required to add/remove any worker node to/from `iscsi_worker` group, then new node can be added or removed as below:

Adding new node to `iscsi_worker` group:

```bash
cray hsm groups members create --id <xname_of_worker_node> iscsi_worker
```

e.g

```bash
# cray hsm groups members create --id  x3000c0s18b0n0 iscsi_worker
```

e.g

> ```text

Reviewer: ````suggestion
```toml
````

Reviewer: This has not been resolved. A number of the example output blocks contain TOML or JSON, but are marked as `text` like this one. They should be marked `toml` or `json`.

[[results]]
URI = "/hsm/v2/groups/iscsi_worker/members/x3000c0s18b0n0"
```

Deleting the worker node from `iscsi_worker` group:

```bash
# cray hsm groups members delete <xname> iscsi_worker
```

e.g.

```text
# cray hsm groups members delete x3000c0s29b0n0 iscsi_worker

ncn-m001:~ # cray hsm groups members delete x3000c0s29b0n0 iscsi_worker
code = 0
message = "deleted 1 entry"
```

Then re-run iSCSI CFS layer on worker nodes using below script:
`/usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py`

> Also, the latest CSM documentation RPM must be installed on the node where the

Reviewer: This needs to go before the script, not after it, because they need this in order to make sure the script is on the node.

> procedure is being performed. See `CFS configuration and component update`

Reviewer: This should be a link to the "Check for documentation updates", not to the HSM groups page.

Author: You mean docs-csm link ?

Reviewer: The title of this link and the link itself are not right. This entire step should basically mirror the format of [this one from the Reboot NCNs procedure](https://github.com/Cray-HPE/docs-csm/blob/0da98916bd98dfbc348ecea24d9f1e5cf59ed7a0/operations/node_management/Reboot_NCNs_manual.md?plain=1#L327-L334)

Reviewer: This has not been resolved.

During fresh install of CSM 1.7.0: `iscsi_worker` HSM group needs to be created before the `Configure management nodes with CFS` step of [CSM 1.7.0 installation](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/install/configure_administrative_access.md#8-configure-management-nodes-with-cfs)

During upgrade of CSM 1.7.0: `iscsi_worker` HSM group needs to be created before ‘management nodes rollout’ step during CSM upgrade using IUF. See [CSM 1.7.0 Upgrade](https://github.com/Cray-HPE/docs-csm/blob/ae9482ae85bc2eede3ac61079ffed5a00fb0a126/operations/iuf/workflows/management_rollout.md)

Post install/upgrade if it is required to add any new worker node to the cluster or need to remove an existing worker node from the cluster, then see the [add/remove worker node](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/operations/node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md#add-worker-storage-or-master-ncns).

Post this, if it is required to add/remove any worker node to/from `iscsi_worker` group, then new node can be added or removed as below:

Adding new node to `iscsi_worker` group:

```bash

> cray hsm groups members create --id <xname of worker node> iscsi_worker

Reviewer 2: Using `_` instead of spaces makes it clearer that this is a single variable, and it keeps the formatting from highlighting the `<` and `>` in red.
```suggestion
cray hsm groups members create --id <xname_of_worker_node> iscsi_worker
```

```

e.g

```bash
# cray hsm groups members create --id  x3000c0s18b0n0 iscsi_worker
```

e.g

```text
[[results]]
URI = "/hsm/v2/groups/iscsi_worker/members/x3000c0s18b0n0"
```

Deleting the worker node from `iscsi_worker` group:

```bash
# cray hsm groups members delete <xname> iscsi_worker
```

e.g.

```text
# cray hsm groups members delete x3000c0s29b0n0 iscsi_worker

ncn-m001:~ # cray hsm groups members delete x3000c0s29b0n0 iscsi_worker
code = 0
message = "deleted 1 entry"
```

> Then re-run CFS on the worker nodes using below script:

Reviewer: You'll also want to state that they need to make sure the latest CSM documentation RPM is installed on the node where they play to run this script. Basically, something like this: https://github.com/Cray-HPE/docs-csm/blob/4802625e1bdc5f17bb76f0e6b40a738242701321/operations/configuration_management/Exporting_and_Importing_CFS_Data.md?plain=1#L11-L12

> Then re-run CFS on the worker nodes using below script:

Reviewer: Once [this PR](https://github.com/Cray-HPE/docs-csm/pull/6126) merges, the script will be available.
The path to the script on the live system will be:
`/usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py`

> TBF

Reviewer: If the [changes for CASMPET-7624/CASMPET-7625](https://github.com/Cray-HPE/csm-config/pull/385) are not merged, then you also need to add information on here about how the Kubernetes labels will not be updated on any workers which are down or which lack iSCSI layers in their CFS configurations. After those changes merge, that isn't necessary, because those changes solve that problem

Author: I see the PR related to  this ticket is merged now.
